Framing for Modbus RTU packets

dhanson358 opened the pull request:
Not sure if this belongs here, or if it would be better to implement this only in a package using nerves-uart, but figured I'd make a PR as a way to open discussion.

Using Modbus packets over uart are a bit weird, because we don't know the length at the time we start receiving.  It comes in the second byte of the response.

This framer will move along incoming data, mark the expected length by reading the second byte, then keep reading until the final size is met, and return that as a complete packet.

fhunleth commented:
Sweet!!! Thanks for sending this up. I am very interested in how people are using the framing API especially for protocols that I don't use (like ModBus).

My thought has been that only having read and write calls in the Nerves.UART API would be too limiting for more complicated protocols like ModBus, and that would lead to a separate ModBus library. I'm not familiar with how ModBus is really used, though. From Wikipedia, it looks like the payload is {address, function, data}. Does it make sense to have an API with calls of the form `function(address, data)`? If so (or something similar), then separate project. If just having read and write makes sense, then let's put it here. Basically, I'm fine with one file and no added dependencies, but if it's more than that, I worry about maintaining it properly.

If we choose separate project, then I'll add a section for related projects so that it's easier for people to discover your ModBus implementation.

dhanson358 commented:
Unfortunately it's not quite so simple.  Modbus breaks down into 2 sub-forms: Modbus TCP and Modbus RTU.  They have the same core payload you describe (function, address, data), but different wrappers (for device ID, crc, etc)...and only RTU is really applicable to a serial connection like nerves-uart supports.

Right now I'm actually using my own branch of another project in coordination with nerves-uart.  The existing project only supported modbus TCP, so I added RTU support and added the dependency on nerves-uart to supply the transport layer.  You can see it here: https://github.com/dhanson358/ex_modbus

My second-choice plan on this PR would be to leave the "Framer" in the ex_modbus library.  We could definitely do it that way and just link from the nerves-uart doc to another option for framing included in the ex_modbus lib.

Thoughts?

fhunleth commented:
It does sound like the best place for all this is the `ex_modbox` library. 

I noticed that you had a fix to `flush` in the commit, could you separate that out and send it up?
